store/tikv/tikvrpc/tikvrpc.go

```diff
@@ -354,3 +372,135 @@ func (resp *Response) GetRegionError() (*errorpb.Error, error) {
+
+// CallRPC launchs a rpc call.
+func CallRPC(ctx context.Context, client tikvpb.TikvClient, req *Request, ch chan<- *TimeoutItem) (*Response, error) {
+	resp := &Response{}
+	resp.Type = req.Type
+	var err error
+	switch req.Type {
+	case CmdGet:
+		resp.Get, err = client.KvGet(ctx, req.Get)
+	case CmdScan:
+		resp.Scan, err = client.KvScan(ctx, req.Scan)
+	case CmdPrewrite:
+		resp.Prewrite, err = client.KvPrewrite(ctx, req.Prewrite)
+	case CmdCommit:
+		resp.Commit, err = client.KvCommit(ctx, req.Commit)
+	case CmdCleanup:
+		resp.Cleanup, err = client.KvCleanup(ctx, req.Cleanup)
+	case CmdBatchGet:
+		resp.BatchGet, err = client.KvBatchGet(ctx, req.BatchGet)
+	case CmdBatchRollback:
+		resp.BatchRollback, err = client.KvBatchRollback(ctx, req.BatchRollback)
+	case CmdScanLock:
+		resp.ScanLock, err = client.KvScanLock(ctx, req.ScanLock)
+	case CmdResolveLock:
+		resp.ResolveLock, err = client.KvResolveLock(ctx, req.ResolveLock)
+	case CmdGC:
+		resp.GC, err = client.KvGC(ctx, req.GC)
+	case CmdDeleteRange:
+		resp.DeleteRange, err = client.KvDeleteRange(ctx, req.DeleteRange)
+	case CmdRawGet:
+		resp.RawGet, err = client.RawGet(ctx, req.RawGet)
+	case CmdRawPut:
+		resp.RawPut, err = client.RawPut(ctx, req.RawPut)
+	case CmdRawDelete:
+		resp.RawDelete, err = client.RawDelete(ctx, req.RawDelete)
+	case CmdRawScan:
+		resp.RawScan, err = client.RawScan(ctx, req.RawScan)
+	case CmdCop:
+		resp.Cop, err = client.Coprocessor(ctx, req.Cop)
+	case CmdCopStream:
+		// Use context to support timeout for grpc streaming client.
+		ctx1, cancel := context.WithCancel(ctx)
+		var stream tikvpb.Tikv_CoprocessorStreamClient
+		stream, err = client.CoprocessorStream(ctx1, req.Cop)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		// Read the first streaming response to get CopStreamResponse.
+		// This can make error handling much easier, because SendReq() retry on
+		// region error automatically.
+		var first *coprocessor.Response
+		first, err = resp.CopStream.Recv()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		resp.CopStream = NewCopStreamResponse(stream, first, cancel, ch)
+	case CmdMvccGetByKey:
+		resp.MvccGetByKey, err = client.MvccGetByKey(ctx, req.MvccGetByKey)
+	case CmdMvccGetByStartTs:
+		resp.MvccGetByStartTS, err = client.MvccGetByStartTs(ctx, req.MvccGetByStartTs)
+	case CmdSplitRegion:
+		resp.SplitRegion, err = client.SplitRegion(ctx, req.SplitRegion)
+	default:
+		return nil, errors.Errorf("invalid request type: %v", req.Type)
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return resp, nil
+}
+
+// TimeoutItem is used to implement grpc stream timeout.
+type TimeoutItem struct {
+	cancel  context.CancelFunc
+	timeout time.Time
+	active  int32 // 0 means inactive
+}
+
+// SetDeadline set deadline for this CopStreamResponse object, it usually called before Recv().
+func (resp *CopStreamResponse) SetDeadline(deadline time.Time) {
+	atomic.StoreInt32(&resp.item.active, 1)
+	resp.item.timeout = deadline
```

Comment:
It's possible that the timeout checker see the item is active, then read the old timeout value, then cancelled the stream.

Reply:
The timeout checker will check `active` flag first, which is a atomic int32.
If the checker see it's active, it will check `timeout`.
So timeout is protected by the atomic `active` flag.